### PR TITLE
[deckhouse] separate monitoring module

### PR DESCRIPTION
**Pull request by yalosev**
Signed-off-by: Yuriy Losev <yuriy.losev@flant.com>

## Description
Moves the deckhouse module to the beginning of the modules in order to avoid problems with CRDs.

Adds the separate deckhouse monitoring module.

## Why do we need it, and what problem does it solve?
The deckhouse module gets stuck and isn't updated when CRDs before it are broken. We can avoid this situation by moving the module to run before the others.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: priority-class
type: chore
summary: Changed the module order from 010 to 001.
---
section: deckhouse
type: chore
summary: Changed the module order from 020 to 002.
---
section: monitoring-deckhouse
type: chore
summary: Separate monitoring rules from the `deckhouse` module to handle CRDs creation order.
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

**Comment by z9r5** [comment]
@yalosev I've set the 1.37.0  milestone. Correct if it is wrong, please.

File: modules/340-monitoring-deckhouse/enabled [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+# Copyright 2021 Flant JSC
```
Comment (z9r5) [inline modules/340-monitoring-deckhouse/enabled:3]:
```suggestion
# Copyright 2022 Flant JSC
```

File: testing/matrix/linter/rules/modules/monitoring.go [hunk 2]
```diff
@@ -46,7 +46,7 @@ func monitoringModuleRule(moduleName, modulePath, moduleNamespace string) errors
 	switch moduleName {
 	// These modules deploy common rules and dashboards to the cluster according to their configurations.
 	// That's why they have custom monitoring templates.
-	case "340-extended-monitoring", "340-monitoring-applications", "030-cloud-provider-yandex":
+	case "340-extended-monitoring", "340-extended-deckhouse", "340-monitoring-applications", "030-cloud-provider-yandex":
```
Comment (nabokihms) [inline testing/matrix/linter/rules/modules/monitoring.go:49]: > 340-extended-deckhouse

Seems like an error

File: modules/340-monitoring-deckhouse/charts/helm_lib/templates/_module_name.tpl [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{{- define "helm_lib_module_camelcase_name" -}}
```
Comment (nabokihms) [inline modules/340-monitoring-deckhouse/charts/helm_lib/templates/_module_name.tpl:1]: Helm lib is a copy, not a symlink? AFIK, it should be a symlink.

File: modules/340-monitoring-deckhouse/.helmignore [hunk 4]
```diff
@@ -0,0 +1,7 @@
+crds
```
Comment (nabokihms) [inline modules/340-monitoring-deckhouse/.helmignore:1]: There are more entries in the helmignore file than needed.

File: modules/340-monitoring-deckhouse/.namespace [hunk 5]
```diff
@@ -0,0 +1 @@
+d8-system
```
Comment (nabokihms) [inline modules/340-monitoring-deckhouse/.namespace:1]: d8-monitoring

File: testing/matrix/linter/rules/modules/modules.go [hunk 6]
```diff
@@ -47,7 +47,7 @@ func moduleLabel(n string) string {
 
 func shouldSkipModule(name string) bool {
 	switch name {
-	case "helm_lib", "400-nginx-ingress", "500-dashboard":
+	case "helm_lib", "400-nginx-ingress", "340-monitoring-deckhouse", "500-dashboard":
```
Comment (nabokihms) [inline testing/matrix/linter/rules/modules/modules.go:50]: Why do we need to ignore the deckhouse monitoring module here?

File: modules/002-deckhouse/templates/webhook-handler/deployment.yaml [hunk 7]
```diff
@@ -3,7 +3,7 @@ cpu: 50m
 memory: 100Mi
 {{- end }}
 
-{{- if (.Values.global.enabledModules | has "vertical-pod-autoscaler-crd") }}
+{{- if and (.Values.global.enabledModules | has "vertical-pod-autoscaler-crd") (.Capabilities.APIVersions.Has "autoscaling.k8s.io/v1/VerticalPodAutoscaler") }}
```
Comment (nabokihms) [inline modules/002-deckhouse/templates/webhook-handler/deployment.yaml:6]:
```suggestion
{{- if and (.Values.global.enabledModules | has "vertical-pod-autoscaler-crd") (.Capabilities.APIVersions.Has "autoscaling.k8s.io/v1/VerticalPodAutoscaler") }}
```
:+1:

**Review by nabokihms — CHANGES_REQUESTED** [review]

File: testing/matrix/linter/rules/modules/modules.go (hunk 6 above)
Comment (yalosev) [reply testing/matrix/linter/rules/modules/modules.go]: Fixed

File: modules/340-monitoring-deckhouse/.namespace (hunk 5 above)
Comment (yalosev) [reply modules/340-monitoring-deckhouse/.namespace]: hm, ok

File: modules/340-monitoring-deckhouse/.helmignore (hunk 4 above)
Comment (yalosev) [reply modules/340-monitoring-deckhouse/.helmignore]: Cleaned

File: modules/340-monitoring-deckhouse/charts/helm_lib/templates/_module_name.tpl (hunk 3 above)
Comment (yalosev) [reply modules/340-monitoring-deckhouse/charts/helm_lib/templates/_module_name.tpl]: Ups, IDE magic %)

Fixed

**Review by nabokihms — APPROVED** [review]

**Pull request merged**